src/spinnaker.cpp

```diff
@@ -0,0 +1,1049 @@
+  : Camera{ .set = ::spinnakercam_set,
+            .get = ::spinnakercam_get,
+            .get_meta = ::spinnakercam_get_meta,
+            .get_shape = ::spinnakercam_get_shape,
+            .start = ::spinnakercam_start,
+            .stop = ::spinnakercam_stop,
+            .execute_trigger = ::spinnakercam_execute_trigger,
+            .get_frame = ::spinnakercam_get_frame,
+  }
+  , camera_(camera)
+  , last_known_settings_{}
+  , frame_id_(0)
+  , started_(false)
+{
+    // Sometimes the camera is still initialized from a previous run.
+    // Ideally, we would error here, but that can make the camera indefinitely
+    // unusable in Acquire, so log and reset instead.
+    if (camera->IsInitialized()) {
+        LOGE("Camera was already initialized. De-initializing it before "
+             "re-initializing.");
+        camera_->DeInit();
+    }
+    camera->Init();
+
+    // Acquire only supports certain values of some node values, so set these
+    // once on initialization before getting or setting any other node values
+    // which may depend on them.
+    set_enum_node(camera_->ExposureAuto, genicam_off);
+    set_enum_node(camera_->ExposureMode, genicam_timed);
+
+    get(&last_known_settings_);
+}
+
+SpinnakerCamera::~SpinnakerCamera()
+{
+    try {
+        stop();
+        const std::scoped_lock lock(lock_);
+        camera_->DeInit();
+    } catch (const std::exception& exc) {
+        LOGE("Exception: %s\n", exc.what());
+    } catch (...) {
+        LOGE("Exception: (unknown)");
+    }
+}
+
+void
+SpinnakerCamera::set(struct CameraProperties* properties)
+{
+    const std::scoped_lock lock(lock_);
+    // Set shape before offset because Spinnaker blocks updates
+    // to offset if the shape is too big.
+    maybe_set_shape(properties->shape);
+    maybe_set_offset(properties->offset);
+    maybe_set_binning(properties->binning);
+    maybe_set_sample_type(properties->pixel_type);
+    // Exposure time can depend on lots of properties, so update it last.
+    maybe_set_exposure_time_us(properties->exposure_time_us);
+    maybe_set_input_trigger_frame_start(properties->input_triggers.frame_start);
+    maybe_set_output_trigger_exposure(properties->output_triggers.exposure);
+}
+
+void
+SpinnakerCamera::maybe_set_exposure_time_us(float target_us)
+{
+    if (target_us != last_known_settings_.exposure_time_us) {
+        set_float_node(camera_->ExposureTime, (double)target_us);
+        // TODO: may not need to actually get from camera because try_camera_set
+        // in runtime/source.c calls get after calling set.
+        last_known_settings_.exposure_time_us = (float)camera_->ExposureTime();
+    }
+}
+
+void
+SpinnakerCamera::maybe_set_binning(uint8_t target)
+{
+    if (target != last_known_settings_.binning) {
+        // Only one of horizontal and vertical may be writable, so explicitly
+        // check each before attempting to write.
+        // TODO: should we error/log if neither is writable?
```

Comment:
I'd prefer not erroring. should log and read back what the camera is set to.

Reply:
Done.